sklearn/tree/_utils.pyx

```diff
@@ -301,3 +303,357 @@ cdef class PriorityHeap:
+        left to right in the array ``array_``. ``array_ptr`` is always
+        less than ``capacity``.
+
+    array_ : WeightedPQueueRecord*
+        The array of priority queue records. The minimum element is on the
+        left at index 0, and the maximum element is on the right at index
+        ``array_ptr-1``.
+    """
+
+    def __cinit__(self, SIZE_t capacity):
+        self.capacity = capacity
+        self.array_ptr = 0
+        self.array_ = <WeightedPQueueRecord*> calloc(capacity,
+                                                     sizeof(WeightedPQueueRecord))
+
+        if self.array_ == NULL:
+            raise MemoryError()
+
+    def __dealloc__(self):
+        free(self.array_)
+
+    cdef bint is_empty(self) nogil:
+        return self.array_ptr <= 0
+
+    cdef SIZE_t size(self) nogil:
+        return self.array_ptr
+
+    cdef int push(self, DOUBLE_t data, DOUBLE_t weight) nogil:
+        """Push record on the array.
+        Returns 0 if successful; -1 on out of memory error.
+        """
+        cdef SIZE_t array_ptr = self.array_ptr
+        cdef WeightedPQueueRecord* array = NULL
+        cdef SIZE_t i
+
+        # Resize if capacity not sufficient
+        if array_ptr >= self.capacity:
+            self.capacity *= 2
+            array = <WeightedPQueueRecord*> realloc(self.array_,
+                                                    self.capacity *
+                                                    sizeof(WeightedPQueueRecord))
+
+            if array == NULL:
+                # no free; __dealloc__ handles that
+                return -1
+            self.array_ = array
+
+        # Put element as last element of array
+        array = self.array_
+        array[array_ptr].data = data
+        array[array_ptr].weight = weight
+
+        # bubble last element up according until it is sorted
+        # in ascending order
+        i = array_ptr
+        while(i != 0 and array[i].data < array[i-1].data):
+            array[i], array[i-1] = array[i-1], array[i]
+            i -= 1
+
+        # Increase element count
+        self.array_ptr = array_ptr + 1
+        return 0
+
+    cdef int remove(self, DOUBLE_t value, DOUBLE_t weight) nogil:
+        """Remove a specific value/weight record from the array.
+        Returns 0 if successful, -1 if record not found."""
+        cdef SIZE_t array_ptr = self.array_ptr
+        cdef WeightedPQueueRecord* array = self.array_
+        cdef SIZE_t idx_to_remove = -1
+        cdef SIZE_t i
+
+        if array_ptr <= 0:
+            return -1
+
+        # find element to remove
+        for i in range(array_ptr):
+            if array[i].data == value and array[i].weight == weight:
+                idx_to_remove = i
+                break
+
+        if idx_to_remove == -1:
+            return -1
+
+        # shift the elements after the removed element
+        # to the left.
+        for i in range(idx_to_remove, array_ptr-1):
+            array[i] = array[i+1]
```

Comment:
Oh I read too fast. You are in fact removing in the middle. What I had in mind does not work then.